### PR TITLE
New PID sliders and changed defaults

### PR DESCRIPTION
**Pull request by haslinghuis**
Changes:

- [x] Fixes: https://github.com/betaflight/betaflight-configurator/issues/2545 
- [x] Fixes: https://github.com/betaflight/betaflight-configurator/issues/2564
- [x] Fixes: https://github.com/betaflight/betaflight-configurator/issues/2567
- [x] Fixes: https://github.com/betaflight/betaflight-configurator/issues/2608 
- [x] Enabling sliders keeps tracks of cached slider settings (don't reset after enable slider button).
- [x] Fixes filter slider calculation
- [x] Fixed resetProfile and added tooltip
- [x] Fixed slider warnings while moving sliders
- [x] Fixed backwards compatibility issues
- [x] Disable FF and DD switches and adjust to sliders
- [x] Fixed legacy behavior for FF and DD switches
- [x] Added default sliders to FC object  

:label:  Depends on: https://github.com/betaflight/betaflight/pull/10919

Screenshot from 2021-09-19 01-13-30

Credits to @spatzengr and @ctzsnooze for providing text for the tooltips and slider labels.
I have worked together with @ctzsnooze at some point.

### CODE REVIEW

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> FAIL
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> PASS
 - assigned to an approver -> PASS
 - approver count at least three -> FAIL

**Comment by haslinghuis** [comment]
SonarCloud go away, I don't care about line length if the code is more structured that way.

**Comment by Asizon** [comment]
Sorry @haslinghuis but we loose all backward compatibility here, we loose one of the sliders for old versions and just stick response slider works.

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> FAIL
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> FAIL
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> FAIL
 - assigned to an approver -> PASS
 - approver count at least three -> FAIL

**Comment by haslinghuis** [comment]
Only need to hide sliders according to firmware now we have different labels and improved tooltips. Also we have to decide what do with new pitch_pi_gain slider.

File: src/js/TuningSliders.js
Comment (Asizon) [inline src/js/TuningSliders.js:186]: This parametre can go outside this Api version condition because is aplied to all versions.

File: src/js/TuningSliders.js
Comment (Asizon) [inline src/js/TuningSliders.js:416]: Same here, is aplied to all versions and can go outside.

**Review by Asizon — CHANGES_REQUESTED** [review]

File: src/js/TuningSliders.js
Comment (haslinghuis) [reply src/js/TuningSliders.js]: Fixed

File: src/js/TuningSliders.js
Comment (haslinghuis) [reply src/js/TuningSliders.js]: Fixed.

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> FAIL
 - assigned to an approver -> PASS
 - approver count at least three -> FAIL

**Comment by haslinghuis** [comment]
Only have to fix the last slider as it doesn't record changes yet.

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> PASS
 - assigned to an approver -> PASS
 - approver count at least three -> FAIL

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> FAIL
 - assigned to an approver -> PASS
 - approver count at least three -> FAIL

**Comment by haslinghuis** [comment]
Tried to implement - reverted now.
```
allPidTuningSliders.on('input mouseup keyup', function() {
```
but it was not working as expected (not showing intermediate values as obvious they need to be calculated in firmware)

**Review by ctzsnooze — APPROVED** [review]

**Comment by spatzengr** [comment]
@haslinghuis , when you switch to RP mode vs. RPY, it turns off the sliders and I can't seem to just turn them on again as it switches back to RPY.

**Comment by haslinghuis** [comment]
Dynamic damping slider doesn't save now.

**Comment by spatzengr** [comment]
oxJd5ZupM1

**Comment by haslinghuis** [comment]
Investigation ongoing:
`dMinSwitch`: somehow `dMinRoll`, `dMinPitch` and `dMinYaw` are not in sync after writing and reading using MSP if sliders are enabled by looking at the values in MSPHelper MSP_PID_ADVANCED and MSP_SET_PID_ADVANCED.

**Comment by spatzengr** [comment]
> manual changing values disables sliders in any case. This as designed. After changing values in RP mode disables sliders anyway as they are sent to firmware and are re-evaluated.

Correct, we need to exclude changes to the Yaw axis values from disabling the sliders when in RP mode.

**Comment by haslinghuis** [comment]
After discussion on slack the feedforward switch and dMinSwitch are enabled by default and cannot be changed.
Still think manual changes in RP mode should disable sliders.

File: src/js/main.js
Comment (mikeller) [inline src/js/main.js:251]: This is probably an edge case, but we should ensure here that the restore can not be executed if the user clicks 'Save & Reboot' in the PID Tuning tab.

File: src/js/tabs/pid_tuning.js
Comment (mikeller) [inline src/js/tabs/pid_tuning.js:2188]: Do we need these?

File: locales/en/messages.json
Comment (haslinghuis) [reply locales/en/messages.json:1667]: Fixed 40 cases

File: src/js/tabs/pid_tuning.js
Comment (haslinghuis) [reply src/js/tabs/pid_tuning.js:1192]: Removed

File: src/js/tabs/pid_tuning.js
Comment (haslinghuis) [reply src/js/tabs/pid_tuning.js:1190]: It keeps track of every item changed. When value of the item is the same as the initial one it removes the keypair. When empty config is same as initial.

File: src/js/main.js
Comment (haslinghuis) [reply src/js/main.js:251]: When clicking on another tab while in pid_tuning and something has changed we have to restore to initial settings to avoid user confusion thinking there was nothing saved.

File: src/js/tabs/pid_tuning.js
Comment (haslinghuis) [reply src/js/tabs/pid_tuning.js:2188]: We are investigating some round errors in GUI. Will cleanup when finished.

File: src/js/tabs/pid_tuning.js
Comment (haslinghuis) [reply src/js/tabs/pid_tuning.js]: Not sure I did this right now.

**Comment by haslinghuis** [comment]
@ctzsnooze enabled the dd and ff switch again.

File: src/js/main.js
Comment (haslinghuis) [reply src/js/main.js:251]: Added condition for restore.

**Comment by haslinghuis** [comment]
1. Removed mouseup event for 4.3 (Thanks to Brian White)
2. Made FF and DD switches read-only and update dynamicaly (spoiler there is a bug at 0 with DD resetting to D or something)
3. Made sliderWarning update on slider moves
4. Things I forgot

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> FAIL
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> FAIL
 - assigned to an approver -> PASS
 - approver count at least three -> FAIL

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> FAIL
 - assigned to an approver -> PASS
 - approver count at least three -> FAIL

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> FAIL
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> FAIL
 - assigned to an approver -> PASS
 - approver count at least three -> FAIL

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> FAIL
 - assigned to an approver -> PASS
 - approver count at least three -> FAIL

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> FAIL
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> FAIL
 - assigned to an approver -> PASS
 - approver count at least three -> FAIL

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> FAIL
 - assigned to an approver -> PASS
 - approver count at least three -> FAIL

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> PASS
 - assigned to an approver -> PASS
 - approver count at least three -> FAIL

**Review by ctzsnooze — APPROVED** [review]

**Review by ctzsnooze — DISMISSED** [review]

File: src/js/tabs/pid_tuning.js
Comment (Asizon) [inline src/js/tabs/pid_tuning.js:519]: This function is never used, can be removed.

**Review by Asizon — CHANGES_REQUESTED** [review]

File: src/js/tabs/pid_tuning.js
Comment (haslinghuis) [reply src/js/tabs/pid_tuning.js]: Revised legacy switches

**Comment by haslinghuis** [comment]
Need to look further at legacy part

**Comment by Asizon** [comment]
Seems that there are no limits for Stick response

image

**Comment by Asizon** [comment]
2.
Non expert mode is not working for old versions, also with expert mode sliders limits are same of non expert mode

File: src/js/tabs/pid_tuning.js
Comment (Asizon) [inline src/js/tabs/pid_tuning.js:510]: All this check for non existent feedforward switch can be removed, also can be removed switch from html.

**Review by Asizon — CHANGES_REQUESTED** [review]

File: src/js/tabs/pid_tuning.js
Comment (haslinghuis) [reply src/js/tabs/pid_tuning.js:509]: Added back non expert mode slider limits and also fixed the negative sliders values on 4.3

**Review by ctzsnooze — APPROVED** [review]

**Comment by ctzsnooze** [comment]
Now the firmware side is merged, we must merge this PR asap or slider users will come un-stuck with 4.3.
If changes are required, to adjust behaviour or fix bugs, those changes will need to be on separate PR's.

**Comment by Asizon** [comment]
@haslinghuis good works looks better now, i have seen few more issues hope that final issues:
- (OLD VERSIONS) Seems that there is working the go to defaults when double click for old versions, but number changes to defaults and sliders remains same position.
- The non expert mode for 4.3 doesnt works as expected, when moving slider this does not reach the end of the sliders marking.

**Review by ctzsnooze — DISMISSED** [review]

**Comment by ctzsnooze** [comment]
I'm sure there are some cosmetic bugs to fix, but the basic functionality is tested and works.  It's essential that we merge this PR promptly to match the firmware, which puts sliders on by default, and they won't work with older Configurator versions.
We can deal with cosmetic issues as they are identified.

**Review by Asizon — DISMISSED** [review]

File: locales/en/messages.json
Comment (mikeller) [inline locales/en/messages.json:205]: Not sure why we are switching from html syntax to xhtml here. Also, adding unrelated change like this to an already big pull request puts a lot of unnecessary burden onto reviewers and should be avoided by putting it into a separate pull request.

File: src/js/tabs/pid_tuning.js
Comment (mikeller) [inline src/js/tabs/pid_tuning.js:1387]: This whole process seems to be unneeded, as `MSP_SET_RESET_CURR_PID` already applies the simplified settings for the PID profile being reset: https://github.com/betaflight/betaflight/blob/415d4db5aa355064c49ae0d8cfe4e8b6081d3c35/src/main/flight/pid.c#L230

File: src/js/tabs/pid_tuning.js
Comment (mikeller) [inline src/js/tabs/pid_tuning.js:1950]: camelCase: `disableSliderOnManualChange()`

**Review by mikeller — CHANGES_REQUESTED** [review]
The doubled up application of the simplified tuning values for `MSP_SET_RESET_CURR_PID` should be removed.
And the unrelated changes to the messages file format should go into a separate pull request.

File: locales/en/messages.json
Comment (haslinghuis) [reply locales/en/messages.json]: Fixed. Will do a sweep in separate PR for double spaces and rename <br /> to <br>

File: src/js/tabs/pid_tuning.js
Comment (haslinghuis) [reply src/js/tabs/pid_tuning.js]: Fixed.

File: src/js/tabs/pid_tuning.js
Comment (haslinghuis) [reply src/js/tabs/pid_tuning.js]: Fixed.

**Comment by haslinghuis** [comment]
@ctzsnooze two questions - put in all kind of scenario's

```
TuningSliders.setExpertMode = function() {
    this.expertMode = isExpertModeEnabled();
    $('#slidersPidsBox, #slidersFilterBox').toggleClass('nonExpertModeSliders', !this.expertMode);
    if (semver.gte(FC.CONFIG.apiVersion, API_VERSION_1_44)) {

        // TODO: reset nonExpertModeSliders - changes after first movement.

        // Question 1: do we need to reset unused sliders as they used for further calculations.

        if (!this.expertMode) {
            // reset values to default
            this.sliderDMaxGain = this.SLIDER_DEFAULT.slider_dmax_gain / 100;
            this.sliderIGain = this.SLIDER_DEFAULT.slider_i_gain / 100;
            this.sliderRollPitchRatio = this.SLIDER_DEFAULT.slider_roll_pitch_ratio / 100;
            this.sliderPitchPIGain = this.SLIDER_DEFAULT.slider_pitch_pi_gain / 100;
            this.sliderMasterMultiplier = this.SLIDER_DEFAULT.slider_master_multiplier / 100;

            // does not update - and is not needed
            // $('#sliderDMaxGain').val(this.sliderDMaxGain).trigger('change');
            // $('#sliderIGain').val(this.sliderIGain).trigger('change');
            // $('#sliderRollPitchRatio').val(this.sliderRollPitchRatio).trigger('change');
            // $('#sliderPitchPIGain').val(this.sliderPitchPIGain).trigger('change');
            // $('#sliderMasterMultiplier').val(this.sliderMasterMultiplier).trigger('change');

            // does not update - and is not needed
            // FC.TUNING_SLIDERS.slider_dmax_gain = Math.round(this.sliderDMaxGain * 20) * 5;
            // FC.TUNING_SLIDERS.slider_i_gain = Math.round(this.sliderIGain * 20) * 5;
            // FC.TUNING_SLIDERS.slider_roll_pitch_ratio = Math.round(this.sliderRollPitchRatio * 20) * 5;
            // FC.TUNING_SLIDERS.slider_pitch_pi_gain = Math.round(this.sliderPitchPIGain * 20) * 5;
            // FC.TUNING_SLIDERS.slider_master_multiplier = Math.round(this.sliderMasterMultiplier * 20) * 5;

            // testing - enable this will update sliders at once but disables sliders on save
            // this.calculateNewPids();
            // this.updateFormPids();
            // this.initPidSlidersPosition();
        }

        // Question 2 - Hide slider input, disable slider changes or hide slider

        // hide the slider input
        // $('#sliderDMaxGain').toggle(this.expertMode);
        // $('#sliderIGain').toggle(this.expertMode);
        // $('#sliderRollPitchRatio').toggle(this.expertMode);
        // $('#sliderPitchPIGain').toggle(this.expertMode);
        // $('#sliderMasterMultiplier').toggle(this.expertMode);

        // or disable slider changes
        // document.getElementById('sliderDMaxGain').disabled = !this.expertMode;
        // document.getElementById('sliderIGain').disabled = !this.expertMode;
        // document.getElementById('sliderRollPitchRatio').disabled = !this.expertMode;
        // document.getElementById('sliderPitchPIGain').disabled = !this.expertMode;
        // document.getElementById('sliderMasterMultiplier').disabled = !this.expertMode;

        // or hide the slider
        $('.tab-pid_tuning .DMaxGainSlider').toggle(this.expertMode);
        $('.tab-pid_tuning .advancedSlider').toggle(this.expertMode);
        $('.tab-pid_tuning .masterSlider').toggle(this.expertMode);

        $('.tab-pid_tuning .legacySlider').hide();
        $('.subtab-pid .nonExpertModeSlidersNote').toggle(!this.pidSlidersUnavailable && !this.expertMode);
        $('.subtab-filter .nonExpertModeSlidersNote').toggle((!this.GyroSliderUnavailable || !this.DTermSliderUnavailable) && !this.expertMode);
    } else {
        const dMinShow = parseInt($('.pid_tuning input[name="dMinRoll"]').val()) !== 0;
        $('.tab-pid_tuning .DMaxGainSlider').toggle(this.expertMode && dMinShow);
        $('.tab-pid_tuning .advancedSlider').hide();
        $('.tab-pid_tuning .DMaxGainSlider').hide();
        $('.tab-pid_tuning .baseSlider').hide();
    }
};
```

**Review by ctzsnooze — DISMISSED** [review]

**Comment by haslinghuis** [comment]
@Asizon for disabled sliders I would like to change the color but it doesn't work
```
        $('.tab-pid_tuning .tuningSlider::-webkit-slider-thumb').css('background', '#eeeeee');

```

File: src/js/tabs/pid_tuning.js
Comment (mikeller) [inline src/js/tabs/pid_tuning.js:1386]: What is the reason that we are executing an immediate save operation after this?

Comment (haslinghuis) [reply src/js/tabs/pid_tuning.js:1386]: Now changing a slider go to another tab and come back to pidtuning it's as from the start (unsaved changes are lost) .
Saving values here because as long settings are not saved configurator thinks it is until the FC reboots (unsaved changes are retained until reboot and then lost - which could lead to surprises in the field).

Doing the refresh resets variables holding user changes to reset the state without saving. So perhaps I could store and retrieve those variables instead.

Comment (haslinghuis) [reply src/js/tabs/pid_tuning.js:1386]: `MSP_SET_RESET_CURR_PID` is not resetting the Gyro Filter.

**Comment by haslinghuis** [comment]
LMM:

Screenshot from 2021-09-28 22-30-52

**Comment by haslinghuis** [comment]
Screenshot from 2021-09-28 22-56-06

**Review by ctzsnooze — DISMISSED** [review]

File: src/js/tabs/pid_tuning.js
Comment (mikeller) [reply src/js/tabs/pid_tuning.js:1386]: The problems that I can see with this are:
- it is unexpected to the user that we save settings outside of clicking 'Save';
- the proposed behaviour is inconsistent for different firmware versions;
- this is even more confusing if a TX is connected at the same time and the user is switching profiles through the TX, triggering saves in configurator.

So I think a better way would be to revert to the old behaviour of just discarding the changes (now probably requiring a restoreInitialSettings` on PID profile changes).

Also, the gyro filter is not part of the PID profile, so there is no reason it should be linked in any way to the PID profile functionality.

Comment (haslinghuis) [reply src/js/tabs/pid_tuning.js]: Found a way ;-)
Have removed the tooltip for the button as it's no longer needed. Now it behaves as before.
`initialSettings` were overwritten on tab `reload,` but now have added another flag to workaround.

**Comment by haslinghuis** [comment]
please wait reviewing as I'm fixing legacy filter range

**Comment by haslinghuis** [comment]
Fixed

**Comment by haslinghuis** [comment]
Finished

**Review by mikeller — APPROVED** [review]
`MSP_SET_RESET_CURR_PID` looks much better now - thanks.

I think the direct access to `document.` should be replaced with JQuery to retain consistency - no reason to mix approaches here. And `condition` should be getting a better name. But these can be done after mergeing, so that we can get wider testing of this.

**Pull request merged**
